src/core/server/saved_objects/migrationsv2/index.ts

```diff
@@ -0,0 +1,891 @@
+    LEGACY_DELETE: (state: LegacyDeleteState) =>
+      Actions.updateAliases(client, state.legacyPreMigrationDoneActions),
+  };
+};
+
+export const next = (
+  client: ElasticsearchClient,
+  transformRawDocs: (rawDocs: SavedObjectsRawDoc[]) => Promise<SavedObjectsRawDoc[]>,
+  state: State
+) => {
+  const delay = <F extends (...args: any) => any>(fn: F): (() => ReturnType<F>) => {
+    return () => {
+      return state.retryDelay > 0
+        ? new Promise((resolve) => setTimeout(resolve, state.retryDelay)).then(fn)
+        : fn();
+    };
+  };
+
+  const map = nextActionMap(client, transformRawDocs);
+
+  if (state.controlState === 'DONE' || state.controlState === 'FATAL') {
+    // Return null if we're in one of the terminating states
+    return null;
+  } else {
+    // Otherwise return the delayed action
+    // We use an explicit cast as otherwise TS infers `(state: never) => ...`
+    // here because state is inferred to be the intersection of all states
+    // instead of the union.
+    const nextAction = map[state.controlState] as (
+      state: State
+    ) => ReturnType<typeof map[AllActionStates]>;
+    return delay(nextAction(state));
+  }
+};
+
+export async function migrationStateMachine({
+  client,
+  kibanaVersion,
+  targetMappings,
+  logger,
+  preMigrationScript,
+  transformRawDocs,
+  migrationVersionPerType,
+  indexPrefix,
+}: {
+  client: ElasticsearchClient;
+  kibanaVersion: string;
+  targetMappings: IndexMapping;
+  preMigrationScript?: string;
+  logger: Logger;
+  serializer: SavedObjectsSerializer;
+  transformRawDocs: (rawDocs: SavedObjectsRawDoc[]) => Promise<SavedObjectsRawDoc[]>;
+  migrationVersionPerType: SavedObjectsMigrationVersion;
+  indexPrefix: string;
+}) {
+  const outdatedDocumentsQuery = {
+    bool: {
+      should: Object.entries(migrationVersionPerType).map(([type, latestVersion]) => ({
+        bool: {
+          must: [
+            { exists: { field: type } },
+            {
+              bool: {
+                must_not: { term: { [`migrationVersion.${type}`]: latestVersion } },
+              },
+            },
+          ],
+        },
+      })),
+    },
+  };
+
+  const indexLogger = logger.get(indexPrefix.slice(1));
+
+  const initialState: State = {
+    indexPrefix,
+    controlState: 'INIT',
+    kibanaVersion,
+    preMigrationScript: Option.fromNullable(preMigrationScript),
+    targetMappings,
+    outdatedDocumentsQuery,
+    retryCount: 0,
+    retryDelay: 0,
+    logs: [],
+  };
+
+  const logStateTransition = (oldState: State, newState: State) => {
+    if (newState.logs.length > oldState.logs.length) {
+      newState.logs
+        .slice(oldState.logs.length)
+        .forEach((log) => indexLogger[log.level](log.message));
+    }
+
+    // @ts-expect-error
+    const { logs, outdatedDocuments, ...logState } = newState;
+    indexLogger.info(`${oldState.controlState} -> ${newState.controlState}: `, logState);
```

Comment:
When things go wrong on Cloud it would be a lot easier to help a customer if we had these logs from the start instead of having to enable debug logging first to see what's happening (and then sometimes the failure goes away after debug logging was enabled).

I think we can eventually lower the log level to debug, but when we ship this I'd prefer to have full visibility

Reply:
I forgot the conversation Josh and I earlier, we discussed maybe dumping all of the state data if something goes wrong. So we could leave out the `logState` meta when we're logging with `info` in realtime. However, we still keep all the state transitions in an array (after removing outdatedDocuments to prevent too much memory consumption). Then if we reach the `FATAL` state we do an error log with all these states including the `logState` metadata so that we have the complete picture of what happened.

Reply:
It makes sense to dump the whole state workflow in the logs when the system fails. That's even one of the major upside of a deterministic state machine.